### DISCUSSION TITLE
Multifile Rules - Global start and end?

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [pmd/pmd#4565] [Q&A]
Is there a way to know, when PMD starts and ends the analysis of all files?
I've read (https://github.com/pmd/pmd/discussions/4508), that it calls the Custom Rules for every file once, so start and end in AbstractRule are called only per file.

**Comment by jsotuyod** [comment]
No, rules do not get notified of global analysis events (notice that multiple rule instances may be created during a single analysis, usually one per analysis thread, so this would not make sense).

You can get around this (in a very hacky way) through a custom `Renderer` however, as these are initialized only once upon analysis start, and rendered once upon analysis end.

What you seem to be trying to do is create a 2-phase analysis, with a first data collection stage during file processing, and an actual analysis phase creating violations upon end. This however would not work as it stands:
- Reporting a violation requires a proper context and node on which to report… not only having stateful rules is a smell, but you would actually need to hold AST nodes, which means leaking the whole tree and tokens. Memory will bloat.
- A hacky approach will work on clean runs, but fail miserably when using the analysis cache, as the data gathered during the first phase would be lost and only modified files be parsed once again.

Beyond finding unused non-private methods or coupling issues, we have so far not found a use case for 2-phase analysis to help us prioritize this. Feel free to chip in if you have another one.

**Reply by oowekyala** [reply]
I just want to point out that there is #3920 that is exactly about that.

**Reply by NilsBaumgartner1994** [reply]
Detecting data clumps is also a 2 phase analysis. Since in the first run we have to get all fields and method parameters. In the second run we then have to Analyse which fields and parameters are the same.

Area: Systems and Software Quality
Live Code Smell Detection of Data Clumps in an Integrated Development Environment 

https://www.scitepress.org/DigitalLibrary/Link.aspx?doi=10.5220/0011727500003464

https://enase.scitevents.org/PreviousAwards.aspx

**Reply by NilsBaumgartner1994** [reply]
The discussion: https://github.com/pmd/pmd/discussions/4406 also would benefit from that
